blog/_posts/2016-10-03-StructuredQueries.md

```diff
@@ -0,0 +1,390 @@
+```
+julia> q = @query select(:src, twice_sepal_length = 2 * sepal_length)
+Query with dummy source src
+
+julia> collect(q, src = iris)
+Tables.Table
+│ Row │ twice_sepal_length │
+├─────┼────────────────────┤
+│ 1   │ 10.2               │
+│ 2   │ 9.8                │
+│ 3   │ 9.4                │
+│ 4   │ 9.2                │
+│ 5   │ 10.0               │
+│ 6   │ 10.8               │
+│ 7   │ 9.2                │
+│ 8   │ 10.0               │
+│ 9   │ 8.8                │
+│ 10  │ 9.8                │
+⋮
+with 140 more rows.
+```
+Whatever the name of the dummy source (minus the `:`) was in the query must be the key in the kwarg passed to `collect`. Otherwise, the method will fail:
+```
+julia> collect(q, tbl = iris)
+ERROR: ArgumentError: Undefined source: tbl. Check spelling in query.
+ in #collect#5(::Array{Any,1}, ::Function, ::StructuredQueries.Query{Symbol}) at /Users/David/.julia/v0.6/StructuredQueries/src/query/collect.jl:23
+ in (::Base.#kw##collect)(::Array{Any,1}, ::Base.#collect, ::StructuredQueries.Query{Symbol}) at ./<missing>:0
+```
+
+## The two problems
+
+Now that we've seen what the SQ query framework itself consists of, we can discuss how such a framework may help to solve the column-indexing and nullable semantics problems.  
+
+### Type-inferability
+
+Recall that the column-indexing problem consists in the inability of type inference to detect the return type of
+
+```
+function f(df::DataFrame)
+    A = df[:A]
+    x = zero(eltype(A))
+    for i in eachindex(A)
+        x += A[i]
+    end
+    return x
+end
+```
+
+What *would* make `f` above amenable to type inference is to pass `A = df[:A]` above to an inner function that executes the loop, for instance
+
+```
+f_inner(A)
+    x = zero(eltype(A))
+    for i in 1:length(A)
+        x += A[i]
+    end
+    return x
+end
+```
+As long as `f_inner` does not get inlined, type inference will run "at" the point at which the body of `f` calls `f_inner` and will have access to the `eltype` of `df[:A]`, since the latter is passed as an argument to `f_inner`.
+
+This strategy of introducing a function barrier also works when one requires multiple columns. For instance, suppose I wanted to generate a new column `C` where `C[i] = g(A[i], B[i])`. The following solution is type-inferable since the type parameters of the zipped iterator `zip(A, B)` reflects the `eltype`s of `A` and `B`:
+
+```
+function f(g, df)
+    A, B = df[:A], df[:B]
+    C = similar(A)
```

Comment:
It probably works in this case, but I would imagine that calling `C = similar(A)` would not work generally since you effectively need to do `C = similar(Core.Inference.return_type(g, (A[1], B[1]))`. Probably not a big deal thoug